src/hooks/room/useJoinRule.ts

```diff
@@ -0,0 +1,30 @@
+/*
+Copyright 2025 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
+Please see LICENSE files in the repository root for full details.
+*/
+
+import { useEffect, useState } from "react";
+import { EventType, type MatrixEvent, type Room, RoomStateEvent, type JoinRule } from "matrix-js-sdk/src/matrix";
+import { type Optional } from "matrix-events-sdk";
+
+import { useTypedEventEmitter } from "../useEventEmitter";
+
+/**
+ * Helper to retrieve the join rules for given room
+ * @param room
+ * @returns the current join rule
+ */
+export function useJoinRule(room?: Room): Optional<JoinRule> {
+    const [topic, setJoinRule] = useState(room?.getJoinRule());
+    useTypedEventEmitter(room?.currentState, RoomStateEvent.Events, (ev: MatrixEvent) => {
+        if (ev.getType() !== EventType.RoomJoinRules) return;
+        setJoinRule(room?.getJoinRule());
+    });
+    useEffect(() => {
+        setJoinRule(room?.getJoinRule());
+    }, [room]);
```

Comment:
Why not use `useRoomState` which already handles the event emitter and other logic for you

Reply:
I think this was due to useRoomState using `RoomStateEvent.Update` which seems to be less reliable in a test environment than `RoomStateEvent.Events`. Looking at it again, it looks like that's just because it's emitted earlier so probably could just make the tests more tolerant.

Reply:
`Update` behaves more expectedly if you get a sync which has N matching state events, e.g. join rule flip-flopping, you don't really want the UI to flash on-off-on-off-on-off as it processes each event in turn. Update only fires at the end of the processing of all state events in the sync with the new state of the room, the new edge the UI should react to, culling all intermediary transitions.